### PR TITLE
feat(client configurations): adding timeout and fallback configurations

### PR DESCRIPTION
**Pull request by yossi-frontegg**
adding timeout and fallback configuraitons in case agent is not responsive or request failed for any reason

### CODE REVIEW

**Review by Omers-Frontegg — APPROVED** [review]

File: src/client-configuration.ts [hunk 1]
```diff
@@ -1,5 +1,11 @@
 import { AxiosInstance } from 'axios';
 import { LoggingClient } from './logging';
+import {
+	FeatureEntitlementsContext,
+	PermissionsEntitlementsContext,
+	RequestContextType,
+	RouteEntitlementsContext
```
Comment (raz-shlomo-frontegg) [inline src/client-configuration.ts:7]: remove unused imports

File: src/entitlements-client-factory.ts [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
 import { OpaQueryClient } from './opa-queries';
 import { ConfigurationInputIsMissingException } from './exceptions/configuration-input-is-missing.exception';
+import { ConfigurationInputIsInvalidException } from './exceptions/configuration-input-is-invalid.exception';
 
 export class EntitlementsClientFactory {
 	public static create(configuration: ClientConfiguration): EntitlementsClient {
 		if (!configuration.pdpHost) {
 			throw new ConfigurationInputIsMissingException('pdpHost is required');
 		}
+		if (configuration.timeout !== undefined && configuration.timeout <= 0) {
```
Comment (raz-shlomo-frontegg) [inline src/entitlements-client-factory.ts:14]: why not use default? i think we should have default timeout (unless debug mode)

File: src/client-configuration.ts [hunk 3]
```diff
@@ -8,4 +14,15 @@ export interface ClientConfiguration {
 		client?: LoggingClient;
 		logResults?: boolean;
 	};
+	timeout?: Milliseconds;
+	fallbackConfiguration?: FallbackConfiguration;
 }
+
+export type Milliseconds = number;
+
+export type FallbackConfiguration = {
```
Comment (raz-shlomo-frontegg) [inline src/client-configuration.ts:23]: suggest to support also functions that returns boolean

File: src/entitlements-client-factory.ts (hunk 2 above)
Comment (yossi-frontegg) [reply src/entitlements-client-factory.ts:14]: There's a default timeout from axios.
If we want to change that, pass a timeout

File: src/entitlements.client.ts [hunk 4]
```diff
@@ -1,33 +1,72 @@
...
+	}
+
+	private async constructFallbackResult(requestContext: RequestContext): Promise<EntitlementsResult> {
+		let fallback: boolean;
+		if (this.fallbackConfiguration instanceof Function) {
+			fallback = await this.fallbackConfiguration(requestContext);
+		} else {
+			const { defaultFallback } = this.fallbackConfiguration;
+			let specificFallback: boolean | undefined;
+			switch (requestContext.type) {
```
Comment (raz-shlomo-frontegg) [inline src/entitlements.client.ts:48]: i would use a map instead of switch case

**Review by raz-shlomo-frontegg — APPROVED** [review]

**Pull request merged**